Design Picker: Open Patterns modal when landing in the editor with Blank Canvas

yansern opened the pull request:
## Changes proposed in this Pull Request
Show the patterns inserter modal when landing in the editor for all the new sites created with Blank Canvas design. To achieve this behavior, we're using the same logic as for the Welcome Tour (see https://github.com/Automattic/wp-calypso/issues/52049). For sites created with Blank Canvas design, the Welcome Guide is displayed after the user interacts with the patterns inserter.

**Demo:** https://cloudup.com/civmwCZ_qhv

The order of events when landing first time in the editor with Blank Canvas is identical to what currently exists in production in this case ([demo](https://cloudup.com/cv6DOanjhon)):
- a new or existing user creates a new site on `/start` so they land in My Home
- the user goes to `https://wordpress.com/page/[SITE_NAME]` to add a new page before accessing the editor previously 
- Page Layout picker is displayed over the editor
- Page Layout picker will continue to be displayed if the page is refreshed without dismissing the modal or inserting a pattern
- After a layout is inserted, the Welcome Guide starts

**Technical changes:**
1. The `wpcom/v2/sites/SITE_ID_HERE/block-editor/nux` API endpoint now returns `variant` property with `blank_canvas_tour` value instead of `tour` for sites created with Blank Canvas design
2. Page Patterns modal shows a custom `title` and `description` when opened using the action type `OPEN_FOR_BLANK_CANVAS`
3. Closing Page Patterns modal (both for page and pattern picking) dispatches `WPCOM_HAS_USED_PATTERNS_MODAL` on the NUX store to prevent opening it again in the Blank Canvas flow.
4. The Welcome Tour for sites created with Blank Canvas is postponed until a pattern is inserted or Page Patterns modal is dismissed

**Notes:**
- Opening of Page Patterns modal is only limited to desktop view.
- In the first 30 minutes after creating a site, when accessing the editor in different browsers (or after clearing localStorage), Page Patterns modal + Welcome Tour will be displayed again (as it currently happens with Welcome Tour)

## Testing instructions

**Setting up:**
1. Apply D60468-code to your sandbox 
2. Run `yarn dev --sync` on `apps/editing-toolkit` on this branch.

**Testing the custom behavior for Blank Canvas when editing homepage**
1. Create a site with Blank Canvas using [/start/with-design-picker](https://calypso.live/start/with-design-picker?branch=add/blank-canvas-auto-open-patterns-inserter-panel) on this branch (locally or with live link)
2. Sandbox the newly created site
3. From the Site Setup checklist, go to Edit Homepage to land in the Editor
4. The Page Pattern modal should be displayed on top of the editor with a custom title ("Choose a Pattern") and a custom description ("Pick a pre-defined layout or continue with a blank page").
5. Refresh the page and the Page Patterns modal should still be displayed over the editor.
6. Once a pattern is inserted or the Page Patterns modal is dismissed, welcome modal is displayed

**Regression testing**
A. Site without Blank Canvas
  1. Create a site with any design **other than** Blank Canvas using [/start/with-design-picker](https://calypso.live/start/with-design-picker?branch=add/blank-canvas-auto-open-patterns-inserter-panel) on this branch (locally or with live link)
  2. Sandbox the newly created site
  3. From the Site Setup checklist, go to Edit Homepage to land in the Editor. Welcome tour is displayed.
  
B. Behavior for a site with Blank Canvas when adding a page: https://github.com/Automattic/wp-calypso/pull/52288#discussion_r631573808  

### Screenshots
<img width="1277" alt="Screenshot 2021-05-12 at 12 36 57" src="https://user-images.githubusercontent.com/14192054/117953648-e78b7800-b31e-11eb-8657-6cd166c372aa.png">

Fixes #49048

yansern committed:
Open patterns inserter panel when site uses blank canvas.

yansern committed:
Open patterns panel when wpcom welcome guide is dismissed.

yansern committed:
Revert patterns code in launch.

yansern committed:
Move over to nux.

yansern commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/class-wp-rest-wpcom-block-editor-nux-status-controller.php:
```diff
@@ -87,10 +87,13 @@ public function get_nux_status() {
 
 		$show_welcome_guide = $this->show_wpcom_welcome_guide( $nux_status );
 
+		$should_open_patterns_panel = get_option( 'current_theme' ) === 'Blank Canvas';
```
Somehow using `get_option( 'is_blank_canvas' )` doesn't work, the property doesn't exist, although the site was created via `/new?flags=gutenboarding/alpha-templates`. For now I'm checking via the template name. Maybe something to investigate?

ciampo replied:
I believe that's because D60468-code hasn't been merged yet.

You'd need to:

1. Apply D60468-code to your sandbox
2. Create a site that uses the Blank Canvas design
3. Add the newly created site to your sandbox
4. At that point, `get_option( 'is_blank_canvas' )` _should_ work

yansern replied:
Oh I see the status is **Accepted** not **Merged**!  That's why!

razvanpapadopol replied:
Updated that diff to use the site option we're now passing when creating the site. See https://github.com/Automattic/wp-calypso/pull/52288/commits/b6b2d670bb7de42499fe437bb9b6f45a514ffd33

yansern commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/open-patterns-panel.js:
```diff
@@ -0,0 +1,41 @@
+// Simulate click that react responds to
+const mouseClickEvents = [ 'mousedown', 'click', 'mouseup' ];
+const simulateMouseClick = function ( element ) {
+	mouseClickEvents.forEach( ( mouseEventType ) =>
+		element.dispatchEvent(
+			new window.MouseEvent( mouseEventType, {
+				view: window,
+				bubbles: true,
+				cancelable: true,
+				buttons: 1,
+			} )
+		)
+	);
+};
```
Note: If you're attempting to add breakpoints to the click simulation codes, bear in mind that React is sensitive to the timing between the chain of events between mousedown/click/mouseup. The inserter panel might open, but somehow, the button's click handler won't work anymore after that. At least that was my experience.

Commit:
Use `is_blank_canvas` site option when creating the site in New Onboarding.

- Export isBlankCanvasDesign utility fn from 'design-picker' package

StefanNieuwenhuis commented on client/landing/gutenboarding/stores/onboard/actions.ts:
```diff
@@ -105,6 +107,7 @@ export function* createSite( {
 			...( anchorFmSpotifyUrl && {
 				anchor_fm_spotify_url: anchorFmSpotifyUrl,
 			} ),
+			is_blank_canvas: isBlankCanvasDesign( selectedDesign ),
```
Are we sure that the `selectedDesign` is always defined at this point? I'm asking this since we're using optional chaining for it at [several](https://github.com/Automattic/wp-calypso/blob/b6b2d670bb7de42499fe437bb9b6f45a514ffd33/client/landing/gutenboarding/stores/onboard/actions.ts#L67) [places](https://github.com/Automattic/wp-calypso/blob/b6b2d670bb7de42499fe437bb9b6f45a514ffd33/client/landing/gutenboarding/stores/onboard/actions.ts#L91) in the code and `isBlankCanvasDesign` doesn't accept `undefined`.

razvanpapadopol replied:
Good catch. Fixed in https://github.com/Automattic/wp-calypso/pull/52288/commits/8bd7634f19438bf7cc5769df1fb2cd905ef2a991

Commit:
set is_blank_canvas site option only if selectedDesign is defined

Commit:
Set correct value for $should_open_patterns_panel

Commit:
Open Page Patterns modal when dismissing the tour for a site created with Blank Canvas

Commit:
Cleanup open-patterns-panel file

Commit:
Use custom title when Page Pattern modal is opened for Blank Canvas

p-jackson commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/class-wp-rest-wpcom-block-editor-nux-status-controller.php:
```diff
@@ -87,10 +87,13 @@ public function get_nux_status() {
-				'show_welcome_guide' => $show_welcome_guide,
-				'variant'            => $variant,
+				'show_welcome_guide'         => $show_welcome_guide,
+				'variant'                    => $variant,
+				'should_open_patterns_panel' => $should_open_patterns_panel,
```
When I made `variant` a string rather than just having a boolean like `show_tour: true` I imagined situations were there might be multiple variants like `variant` could be "modal" or "tour" or "tour v2" or "podcast tour".
I wonder whether this is like a variant too? Like `variant === "blank-canvas-tour"`? Maybe not, I haven't thought about it very much, just raising it as an option.

razvanpapadopol replied:
Thanks for pointing this out! I'm also not 100% sure but for now, it seems to make the logic a bit more readable.
Fixed in dc61f62.

p-jackson commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-launch.js:
```diff
@@ -68,6 +68,12 @@ function WelcomeTourFrame() {
+	const { setOpenState } = useDispatch( 'automattic/starter-page-layouts' );
+
+	const { shouldOpenPatternsPanel, isManuallyOpened } = useSelect( ( select ) => ( {
+		shouldOpenPatternsPanel: select( 'automattic/wpcom-welcome-guide' ).shouldOpenPatternsPanel(),
```
Thinking this sort of thing can be renamed now?

```suggestion
		shouldOpenPagePatternsModal: select( 'automattic/wpcom-welcome-guide' ).shouldOpenPagePatternsModal(),
```

p-jackson commented on packages/page-pattern-modal/src/components/page-pattern-modal.tsx:
```diff
@@ -313,7 +314,7 @@ class PagePatternModal extends Component< PagePatternModalProps, PagePatternModa
 							className="page-pattern-modal__heading"
 						>
-							{ __( 'Add a page', __i18n_text_domain__ ) }
+							{ this.props.title || __( 'Add a page', __i18n_text_domain__ ) }
```
I think there was some discussion in the p2 about changing the description as well as the title? Feels like "start with a blank page" doesn't feel right since they just asked to do that during onboarding. Maybe "continue with blank page"?

razvanpapadopol replied:
Yes, you're right. Now I see in p9Jlb4-2qt-p2#comment-3874

> The subtitle also needs adjustment

@ollierozdarz, WDYT? Should we go with "Pick a pre-defined layout or continue with a blank page" as subtitle?

ollierozdarz replied:
Apologies @razvanpapadopol I missed this! Great suggestion @p-jackson , **continue** here makes complete sense and feels like a good progression.

Let's roll with the suggestion, "Pick a pre-defined layout or continue with a blank page".

razvanpapadopol replied:
Done! Here's how it looks together with the smaller font for the title as suggested by @olaolusoga 

<img width="1277" alt="Screenshot 2021-05-12 at 12 36 57" src="https://user-images.githubusercontent.com/14192054/117954119-5f59a280-b31f-11eb-8b60-8f8b55803d8e.png">

p-jackson commented:
The technical approach here is nice and clear given the flow you're going for 👍 Made that comment about the `variant` API param, but I honestly could go either way on that.

Some UX things:
- Opening a modal after clicking the [x] does feel a little strange. It's so abrupt I was almost expecting some sort of popup tooltip to explain what just happened.
- After landing in the editor and dismissing the tour the user can make all sorts of editors. This is how the tour was designed and if they've got a big enough screen there's no reason why not too. After the dismissing tour they could click any button other than [x] and they'll lose all their work without any warning. https://d.pr/v/hQ779T
- Will the `gutenboarding/alpha-templates` flag ever be active for existing users or are we only going to trial this on new users? Because if I do this as an existing user (so the `FORCE_SHOW_WPCOM_WELCOME_GUIDE` behaviour wouldn't be there) then the tour will never open, which means the modal will never appear when existing users choose the blank canvas design. Maybe this is by design and we only want this to help new users.

What I'm wondering is whether the modal should open immediately after landing in the page after onboarding, and then when the modal is closed the tour opens?

razvanpapadopol commented:
> Made that comment about the `variant` API param, but I honestly could go either way on that.

Thanks for pointing that out! We'll be looking into it.

> - Opening a modal after clicking the [x] does feel a little strange. It's so abrupt I was almost expecting some sort of popup tooltip to explain what just happened.
> - After landing in the editor and dismissing the tour the user can make all sorts of editors.
>
> What I'm wondering is whether the modal should open immediately after landing in the page after onboarding, and then when the modal is closed the tour opens?

Yes, we'll reverse the order as @olaolusoga mentioned in points 2 and 4 of p9Jlb4-2qt-p2#comment-3874

> * Will the `gutenboarding/alpha-templates` flag ever be active for existing users or are we only going to trial this on new users? Because if I do this as an existing user (so the `FORCE_SHOW_WPCOM_WELCOME_GUIDE` behaviour wouldn't be there) then the tour will never open, which means the modal will never appear when existing users choose the blank canvas design. Maybe this is by design and we only want this to help new users.

We'll be rolling out Blank Canvas to all users. The testing instructions are like this until https://github.com/Automattic/wp-calypso/issues/52049 is done.

p-jackson commented:
> We'll be rolling out Blank Canvas to all users. The testing instructions are like this until #52049 is done.

Ah of course! I should have known that :)

Commit:
Introduce blank-canvas-tour variation that opens Patterns Picker.
- show Patterns picker before Welcome Tour for sites created with Blank Canvas design
- remove should_open_patterns_panel from nux API response
- remove shouldOpenPatternsPanel selector
- remove shouldOpenPatternsPanelReducer reducer

Commit:
Cleanup tour-launch.js

Commit:
Support adding is_blank_canvas site option in /start

Commit:
Add custom description to patterns modal

Commit:
Reduce the font-size for patterns modal title

alshakero commented on apps/editing-toolkit/editing-toolkit-plugin/starter-page-templates/page-patterns-plugin.tsx:
```diff
@@ -24,13 +25,20 @@ export function PagePatternsPlugin( props: PagePatternsPluginProps ): JSX.Elemen
 			areTipsEnabled: select( 'core/nux' )
 				? ( select( 'core/nux' ).areTipsEnabled() as boolean )
 				: false, // Gutenberg 7.1.0 or lower
+			...( isPatternPicker() && {
```
What happens when `isPatternPicker` returns false? I think spreading `false` would throw.

razvanpapadopol replied:
It's not throwing. Actually, this is a pattern we use quite intensively. See https://github.com/Automattic/wp-calypso/blob/trunk/client/landing/gutenboarding/stores/onboard/actions.ts#L91-L109

alshakero replied:
Yes, we use it for object or undefined|null. But I haven't seen it for booleans.

alshakero replied:
This throws, you can copy it to browser console:

```js
const bool = false;
console.log(...(bool && {name: 'Omar'}))
```

alshakero replied:
It seems even TS isn't happy 
![image](https://user-images.githubusercontent.com/17054134/117990954-e9186880-b33d-11eb-8caa-77f11c0752ad.png)

razvanpapadopol replied:
The example is going to throw an error also if `bool` is `undefined`. 
Try wrapping the spread in an object: `console.log( { ...(bool && {name: 'Omar'}) } )`

alshakero replied:
That makes sense! I tested that and it works. I'm confused as to why TS is complaining now.

razvanpapadopol replied:
I guess in your local environment, TS picks `isPatternPicker()` as 100% falsy. It may be falsy at times but not always.

This may shed some light:

> Spreading in a value that is sometimes a non-object is idiomatic and not a problem.
> Spreading in a value that is always a non-object indicates a logic error in your code.

from: https://github.com/microsoft/TypeScript/issues/37491#issuecomment-601741064

p-jackson replied:
Type error went away for me once I restarted the TS server 🤷‍♂️
When I hover over `isPatternPicker` it says `() => boolean`, which is correct and shouldn't cause problems with spreading. Before I restarted the TS server it said something like `() => unknown` which is probably what was causing the problem.

alshakero commented on client/lib/signup/step-actions/index.js:
```diff
@@ -55,6 +55,7 @@ import flows from 'calypso/signup/config/flows';
 import steps, { isDomainStepSkippable } from 'calypso/signup/config/steps';
 import { isEligibleForPageBuilder, shouldEnterPageBuilder } from 'calypso/lib/signup/page-builder';
 import { fetchSitesAndUser } from 'calypso/lib/signup/step-actions/fetch-sites-and-user';
+import isBlankCanvasDesign from '@automattic/design-picker';
```
This breaks signing up because it imports the DesignPicker component. 

Another day of appreciating TypeScript.

```suggestion
import { isBlankCanvasDesign } from '@automattic/design-picker';
```

razvanpapadopol replied:
Good catch! Yeah, our world would be a really scary place without TypeScript 😬

alshakero reviewed (changes requested):
Reviewed and tested! LGTM. 

But requesting changes because this introduced a bug to sign up that I explain below. 

I tested by fixing it manually and all works ✨

alshakero committed:
Fix isBlankCanvasDesign import

alshakero reviewed (approved):
I pushed the fix to save the back and forth. So approving.

Commit:
Keep the bigger font-size for PagePatternModal default title

p-jackson commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/block-editor-nux.js:
```diff
@@ -41,6 +51,14 @@ registerPlugin( 'wpcom-block-editor-nux', {
 
+		// Open patterns panel before Welcome Tour if necessary (e.g. when using Blank Canvas theme)
+		// Do this only when Welcome Tour is not manually opened.
+		if ( variant === 'blank-canvas-tour' && ! isManuallyOpened ) {
```
I think this logic is causing the page pattern modal to open **twice** if the user doesn't follow the site setup checklist.

1. Create a new site at `/start/with-design-picker`
2. Choose blank canvas
3. After landing on My Home, **don't choose edit homepage from checklist**
4. Create a new page another way e.g. go to `/pages` list and click "Add new page"
5. Page pattern modal opens as expected, make a selection
6. **Notice that the page pattern modal opens a second time**

I think maybe the issue is that there's an assumption that the first call to `/block-editor/nux` is going to be for the blank homepage, but that's not necessarily the case.

p-jackson commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-launch.js:
```diff
@@ -21,7 +21,7 @@ function LaunchWpcomWelcomeTour() {
 	const { show, isNewPageLayoutModalOpen, isManuallyOpened } = useSelect( ( select ) => ( {
 		show: select( 'automattic/wpcom-welcome-guide' ).isWelcomeGuideShown(),
-		// Handle the case where the new page layout modal is initialized and open
+		// Handle the case where the new page pattern modal is initialized and open
```
Thanks ❤️

p-jackson commented on client/state/signup/dependency-store/schema.js:
```diff
@@ -34,6 +34,7 @@ export const dependencyStoreSchema = {
 				headings: { type: 'string' },
 				base: { type: 'string' },
 			},
+			slug: { type: 'string' },
```
I don't see where this new `slug` bit is used in the PR, may have missed something.

razvanpapadopol replied:
When we're calling `isBlankCanvasDesign` function [here](https://github.com/Automattic/wp-calypso/pull/52288/files#diff-0250d6497eaaa43f73751fc190ae92a7d78e96be99f83f856ca3e100cd255d06R110), it expects a `Design` object that contains `slug`. TypeScript could've helped here as well 🙈

p-jackson reviewed (dismissed):
Code's looking good, and I tested various flows to see if I could break it. The issue where the user opens the editor without using the checklist (see below) is the only problem I found.

From watching user testing sessions it seems plenty of folks will skip over the setup checklist entirely and do their own thing, so I think it's fairly expected that entering the editor via a non-checklist route is pretty expected.

Commit:
Use block-editor-nux.js just to open Patterns Picker modal for Blank Canvas

Commit:
Update NUX store to replace 'blank-canvas-tour' with 'tour' after Patterns Modal has been used once and not only displayed.
- introduce BLANK_CANVAS_VARIANT  and DEFAULT_VARIANT contants
- update tests

Commit:
Dispatch setUsedPageOrPatternsModal action on NUX store every time when closing Patterns Modal regardless of the variant.
- change setOpenState prop of PagePatternModal to onClose and handling store update in the parent component.

alshakero reviewed (approved):
Reviewed and tested against [these steps](https://github.com/Automattic/wp-calypso/pull/52288#discussion_r631573808). LGTM!

p-jackson reviewed (commented):
Bug fixed 👍

The flow can still break if the user wanders off the extact track we're expecting them to follow e.g. make new blank site, add new page from `/pages`, go back to My Home, use the "update homepage" button in the checklist, the page layout modal doesn't appear.
But perhaps this is ok for now? Depends on the goals of your project.

If you did want to fix this (maybe in the future), I'm actually wondering now whether `/block-editor/nux` was never the right endpoint to modify. I may have put you wrong by suggesting that the `variant` prop could be extended implement this behaviour.
In order to have this behaviour only happen when updating the homepage I think the logic in PHP would be more like `get_option( 'is_blank_canvas' ) && is_editing_page_on_front()`. But the `/block-editor/nux` endpoint is heavily cached (intentionally), and the `is_editing_page_on_front()` check isn't suitable for caching.

Maybe that means a seperate endpoint? Or maybe there's a way to do the `get_option( 'is_blank_canvas' ) && is_editing_page_on_front()` check in JS?

simison commented:
> But perhaps this is ok for now? Depends on the goals of your project.

I think it's fine for this not to be perfect on first round, and easy enough to do follow-ups.

razvanpapadopol commented:
> I think it's fine for this not to be perfect on first round, and easy enough to do follow-ups.

Thanks for the confirmation, @simison! Yes, the goal here was that all users that have started with Blank Canvas would interact at least once with the Page layout/patterns modal when they land in the editor and we're doing exactly that without being too annoying (see: https://github.com/Automattic/wp-calypso/pull/52288#discussion_r631573808)

> I may have put you wrong by suggesting that the `variant` prop could be extended implement this behaviour.

Not at all! Considering this behavior as part of the Welcome tour for Blank Canvas actually makes a lot of sense and it's also suggested in p9Jlb4-2qt-p2 by @ollierozdarz 

Anyway, we should keep in mind this is just i1 so there will be more improvements to come 🚀 
Thanks everyone for the feedback and help with this PR! ❤️

Commit:
update blank canvas site option

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/5919380

Thank you @yansern for including a screenshot in the description! This is really helpful for our translators.

yansern commented:
> Thank you @yansern for including a screenshot in the description! This is really helpful for our translators.

No, thanks @razvanpapadopol for the updated screenshot. ;)

razvanpapadopol commented:
ETK update deployed in D61483-code

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/5919380).
